### ISSUE TITLE
DH key exchange implementation is slow (pow)

### ISSUE THREAD

**Issue by iwanb** [ronf/asyncssh#472]
The Diffie-Hellman implementation is slow because of the use of `pow` on such big numbers: https://github.com/ronf/asyncssh/blob/develop/asyncssh/kex_dh.py#L199=

I expect replacing the implementation with the one from `cryptography` would speed things up (and be safer): https://cryptography.io/en/latest/hazmat/primitives/asymmetric/dh/?highlight=diffie-hellman#cryptography.hazmat.primitives.asymmetric.dh.DHParameters

**Comment by ronf** [comment]
Thanks for pointing this out! This is one of the oldest pieces of code in AsyncSSH, dating back about 9 years to before I was using the cryptography package.

I had meant to eventually replace this with a wrapper around the DH class in cryptography, but I've never gotten around to it and I didn't really expect there to be much of a speed difference, since pow() was a single call that should be implemented in C.

Now that I've switched things over, though, I was quite surprised to see around an 11-13x speedup! On group1, I'm seeing the time to do a DH exchange drop from 12 msec to under 1 msec. For group14, it goes from around 77 msec to 6 msec, and for group18, it goes from 3.9 sec to about 340 msec...

If you'd like to give this a try, it's available in commit d59ba85 in the "develop" branch.

**Comment by iwanb** [comment]
Great, on my setup it went from:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       16   13.228    0.827   13.228    0.827 {built-in method builtins.pow}
```

to I can't find the key exchange in the profile :)

Thanks a lot!

**Comment by ronf** [comment]
My pleasure! Thanks for the report and confirmation that it works for you. This will be included in the next AsyncSSH release.

**Comment by ronf** [comment]
This change is now available in the AsyncSSH 2.11.0 release.
